src/elfpy/markets.py

```diff
@@ -67,67 +95,117 @@ def __init__(
+                f'markets.swap: ERROR: agent_action.action_type should be an allowed action for the model={self.pricing_model.model_name()}, not {agent_action.action_type}!'
             )
+        # TODO: check the desired amount is feasible, otherwise return descriptive error
+        # update market variables which may have changed since the user action was created
+        agent_action.update_market_variables(market=self)
+        agent_action.print_description_string()
+
         # for each position, specify how to forumulate trade and then execute
-        if user_action.action_type == "open_long":  # buy to open long
-            user_action.direction = "out"  # calcOutGivenIn
-            user_action.token_out = "pt"  # buy unknown PT with known base
-            market_deltas, wallet_deltas = self._open_long(user_action)
-        elif user_action.action_type == "close_long":  # sell to close long
-            user_action.direction = "out"  # calcOutGivenIn
-            user_action.token_out = "base"  # sell known PT for unkonwn base
-            market_deltas, wallet_deltas = self._close_long(user_action)
-        elif user_action.action_type == "open_short":  # sell PT to open short
-            user_action.direction = "out"  # calcOutGivenIn
-            user_action.token_out = "base"  # sell known PT for unknown base
-            market_deltas, wallet_deltas = self._open_short(user_action)
-        elif user_action.action_type == "close_short":  # buy PT to close short
-            user_action.direction = "in"  # calcInGivenOut
-            user_action.token_in = "base"  # buy known PT for unknown base
-            market_deltas, wallet_deltas = self._close_short(user_action)
+        if agent_action.action_type == "open_long":  # buy to open long
+            agent_action.direction = "out"  # calcOutGivenIn
+            agent_action.token_out = "pt"  # sell known base for unknown PT 
+            market_deltas, agent_deltas = self._open_long(agent_action)
+        elif agent_action.action_type == "close_long":  # sell to close long
+            agent_action.direction = "out"  # calcOutGivenIn
+            agent_action.token_out = "base"  # sell known PT for unknown base
+            market_deltas, agent_deltas = self._close_long(agent_action)
+        elif agent_action.action_type == "open_short":  # sell PT to open short
+            agent_action.direction = "in"  # calcOutGivenIn
+            agent_action.token_out = "pt"  # sell known PT for unknown base
+            market_deltas, agent_deltas = self._open_short(agent_action)
+        elif agent_action.action_type == "close_short":  # buy PT to close short
+            agent_action.direction = "in"  # calcOutGivenIn
+            agent_action.token_in = "pt"  # buy known PT for unknown base
+            market_deltas, agent_deltas = self._close_short(agent_action)
+        elif agent_action.action_type == "add_liquidity":
+            # pricing model computes new market deltas
```

Comment:
I think it would. make more sense for these comments to live inside the function docstring for the `_add_liquidity` class.